Proposal: Default values for function arguments (#7521)
carbon-language/carbon-lang: proposals/p007521-default-values-for-function-arguments.md

    Exploration of _struct-pattern_ changes and an example of a match _alternative-pattern_ in
    generic code where default values are useful.
-   [2026-07-27](https://docs.google.com/document/d/1mjllGO3ZCL4qGt9uJHUtcxKoHAGEY7Y999ie4EtBWB8/edit?tab=t.k8f9m2vciixb#heading=h.6aq96lu1p4kx)
    Further discussion of semantic versus syntactic defaults support, and some feedback on an early
    draft of this proposal to consider downscoping it back to function default values only.

#### Discord Discussions

-   [2026-07-24](https://discord.com/channels/655572317891461132/748959784815951963/1530303233942229053)
    Some questions and answers around the match process for patterns with defaults, as well as what
    to do with implicit conversions and defaults in structures.

#### Leads Questions

The work on this proposal exposed a few questions for Carbon Leads, enumerated here:

-   [#7529](https://github.com/carbon-language/carbon-lang/issues/7529) Do nested patterns with
    entirely specified default values also need to specify a default value to accept callee elision?
-   [#7530](https://github.com/carbon-language/carbon-lang/issues/7530) Which de-structured name

>     binding form should accept defaults?

Reviewer: I think you can compress this some, see suggestion.

The leads questions you're asking to factor out decisions in the proposal I'd probably reference below in the proposal when you get to that point rather than front-loading them here.

```suggestion
-   Carbon's [C++ interoperability goals](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
-   [C++ default arguments](https://en.cppreference.com/cpp/language/default_arguments)
-   [p5164](/proposals/p005164-updates-to-pattern-matching-for-objects.md) affirmed that
    `var` must declare a single complete object.
-   [p1084](/proposals/p001084-generics-details-9-forward-declarations.md) in the associated
    [PR](https://github.com/carbon-language/carbon-lang/pull/1084) added support for the
    `where _` syntax when redeclaring constants in generics.

### Vocabulary

For the scope of this document we define certain terms as follows.

-   **Parameters** are the bindings between names and values used in a function declaration,
    declared in Carbon using its pattern-matching system.
-   **Arguments** are the actual values passed by a caller of the function at runtime, and are the
    scrutinee during pattern matching against the function parameter pattern.
-   **Default Argument Values**, sometimes shortened to default arguments or even defaults,
    are values for function parameters defined in the function declaration that will be provided
    by the compiler at the function call site when those arguments are not provided by the
    developer.

### Meeting Discussions

We have discussed this proposal on a few occasions at the Carbon open meetings. Links to the
minutes for each date and a short summary of the discussions follow:

-   [2026-07-16](https://docs.google.com/document/d/1mjllGO3ZCL4qGt9uJHUtcxKoHAGEY7Y999ie4EtBWB8/edit?tab=t.k8f9m2vciixb#heading=h.6n4qdhlv19kv):
    Suggestion to broaden the scope to _tuple-pattern_ and _struct-pattern_, discussion of semantic
    and syntactic parameters, proposal of leads questions.
    [#7529](https://github.com/carbon-language/carbon-lang/issues/7529) and
    [#7530](https://github.com/carbon-language/carbon-lang/issues/7530)
-   [2026-07-20](https://docs.google.com/document/d/1mjllGO3ZCL4qGt9uJHUtcxKoHAGEY7Y999ie4EtBWB8/edit?tab=t.k8f9m2vciixb#heading=h.dyf1sdo924kx):
    Exploration of _struct-pattern_ changes and an example of a match _alternative-pattern_ in
    generic code where default values are useful.
-   [2026-07-27](https://docs.google.com/document/d/1mjllGO3ZCL4qGt9uJHUtcxKoHAGEY7Y999ie4EtBWB8/edit?tab=t.k8f9m2vciixb#heading=h.6aq96lu1p4kx)
    Further discussion of semantic versus syntactic defaults support, and some feedback on an early
    draft of this proposal to consider downscoping it back to function default values only.

### Discord Discussions

-   [2026-07-24](https://discord.com/channels/655572317891461132/748959784815951963/1530303233942229053)
    Some questions and answers around the match process for patterns with defaults, as well as what
    to do with implicit conversions and defaults in structures.
```

Author: Done.

## Proposal

In Carbon, a function parameter list is actually a
[_tuple-pattern_](/docs/design/pattern_matching.md#tuple-patterns), and an argument list is actually
a tuple literal expression. When you call a function, the argument tuple is matched against the
parameter pattern. Like any _tuple-pattern_, the parameter list can contain arbitrarily complex
subpatterns, not just name bindings. So we need to design how they fit into the pattern syntax, and

> how they affect pattern matching.

Reviewer 2: This seems like it belongs in "Background" or "Problem", because these are pre-existing facts that are relevant to the proposal, not changes that the proposal is making.

Author: Moved and merged with a paragraph in the problem statement.

Because developers can nest patterns arbitrarily, and a
[_struct-pattern_](/docs/design/pattern_matching.md#struct-patterns) is a natural pattern to nest

> in a _tuple-pattern_, we also proposing adding support for defaults to this pattern as well.

Reviewer 2: This seems like it belongs in the Rationale or Alternatives Considered, because it's not saying what the proposal is, it's explaining why the proposal is the way it is.

Author: Merged with existing copy in the alternatives considered section.

Default values are only allowed in function declarations for now. This is an intentional
limitation of the scope of this proposal for expediency, and nothing proposed here is intended to

> apply, in the long term, to function declarations exclusively.

Reviewer 2: It feels weird to spend a paragraph explaining where you are and aren't allowed to do the thing, before you explain what the thing actually is (in the next pararaph).

Author: Moved to further down in the same section.

> Following any subpattern in either _struct-pattern_ or _tuple-pattern_, the programmer can declare

Reviewer 2: I think we want to say that defaults are attached to the _elements_ of struct/tuple patterns (meaning the comma-separated sub-units of those patterns).

```suggestion
Following any element of either _struct-pattern_ or _tuple-pattern_, the programmer can declare
```

Note that this is both broader and narrower than "subpattern": given `{.x = x: i32, .y = (a: i32, b: i32)}`, `.x = x: i32` is an element of the struct pattern but not a subpattern of it (because it's not a pattern at all), and `a: i32` is a subpattern of the struct pattern, but not an element of it (because "subpattern of" is transitive, but "element of" is not).

Author: Ah, I mistakenly thought `.x = x: i32` is also a pattern. Corrected.

an optional default value with the addition of an assignment symbol `=` followed by a

> _value-expression_. Note that Defaults cannot appear in `var` patterns, because the

Reviewer: ```suggestion
_value-expression_. Note that these defaults cannot appear in `var` patterns, because the
```

Author: Done.

initialization of a `var` pattern must be written in a single place, not interleaved with other
code.

In addition, we propose adding limited interoperation support with C++ for specifying default
values for function arguments. Most of the syntax and rules are designed to mirror that of function
argument default values in C++, to ease "tech transfer" for C++ developers to Carbon as well as
possibly reduce interoperability implementation costs.

For now, default value expressions must be concrete constant expressions.

Here's a simple example declaring default values for parameters in a function definition:

```carbon
import Core library "io";

fn DefaultThree(x: i32 = 3) -> i32 {
  return x;
}

fn Main() {
  Core.Print(DefaultThree());  // `3`
  Core.Print(DefaultThree(7));  // `7`
}
```

> ## Out of Scope

Reviewer 2: This isn't one of the conventional top-level sections of a proposal. I'm not sure I see the need to have it at all, but if you keep it, it should probably go in the "Problem" section, since it's basically a list of problems we're not proposing to solve.

Author: Moved and reparented to problems.

element within the _tuple-pattern_ provides a default value, the developers must provide default
values for every element to the right of that element, or the compiler will issue a diagnostic.

#### Right-To-Left Value Elision

Default values for _tuple-pattern_ elements work similarly to C++ function argument defaults in that
the scrutinee may only elide arguments in right-to-left fashion, avoiding the ambiguity of the\
matching the scrutinee-provided values to the elided parameters in unspecified order. For example:

```carbon
fn TwoDefaults(x: i32 = 0, y: i32 = 1);

TwoDefaults();  // OK, x = 0, y = 1
TwoDefauts(1);  // OK, x = 1, y = 1
TwoDefaults(, 3);  // ERROR, value not specified for first argument
```

Note that the right-to-left elision matching process happens at the individual tuple level, meaning
that nested _tuple-pattern_ scrutinees may also elide values in right-to-left order without being
subject to the constraint of where they are nested in a higher-level containing _tuple-pattern_.
For example the following is valid:

```carbon
fn Nested(a: i32, (b: i32, c: i32 = 1), d: i32);

Nested(1, (2), 3);  // OK, a = 1, b = 2, c = 1, d = 3
```

> #### Pattern Matching

Reviewer: We use the Google documentation style guide generally which suggests sentence-case for titles and headings: https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

Author: Fixed everywhere.

During the arity check phase of tuple pattern matching, we can any missing elements to the
scrutinee, drawing from the extended type of the defaults provided in the _tuple-pattern_.
Any type or other match issues created by these additions will be caught by further stages in the
match process.

### Struct Patterns

> We also extend default values to name bindings inside of the _struct-pattern_. Since the

Reviewer 2: ```suggestion
We also extend default values to elements of a _struct-pattern_. Since the
```

Author: Done.

syntax is a bit more complex, particularly in the case of the long form, we have a leads question
[#7530](https://github.com/carbon-language/carbon-lang/issues/7530) about this.

The current proposed syntax for both short and long form is to use the same optional assignment `=`
and value pair as proposed for _tuple-pattern_, for example:

```carbon
// The "long" form, allowing member and binding names to differ
fn Long(index: i32 = 0, {.key = k: i32 = 0, .value = v: i32 = 0} = {});

// The "short" form, keeping member and binding names the same
let Short(index: i32 = 0, {key: i32 = 0, value: i32 = 0} = {});
```

There's a pending leads question [#7529](https://github.com/carbon-language/carbon-lang/issues/7529)

> about the trailing `= {}` on both of these examples.

Reviewer: Should this be moved into the next section? Or cited there?

Author: Moved.

#### Out-Of-Order Elision

In a _struct-pattern_, the right-to-left elision requirements don't apply, as each member has an

> associated name, thus resolving the order ambiguity. This means that any struct member binding that

Reviewer 2: ```suggestion
associated name, thus resolving the order ambiguity. This means that any struct element that
```

Author: Done.

has a provided default may be elided, including all members and the entire structure itself, if
defaults are specified for all named members.

```carbon
Long();  // OK, index = 0, key = 0, value = 0

> Short(1, {.value = 3});  // OK, index = 1, k = 0, v = 3

Reviewer: ```suggestion
Short(1, {.value = 3});  // OK, index = 1, key = 0, value = 3
```

Author: Done.

        -   [Right-To-Left Value Elision](#right-to-left-value-elision)
        -   [Pattern Matching](#pattern-matching)
    -   [Struct Patterns](#struct-patterns)
        -   [Out-Of-Order Elision](#out-of-order-elision)
        -   [Interaction Between Defaults and Unspecified Structure Members](#interaction-between-defaults-and-unspecified-structure-members)
        -   [Pattern Matching](#pattern-matching-1)
    -   [Nesting Parameter Defaults](#nesting-parameter-defaults)
    -   [Interaction with Function Overloading](#interaction-with-function-overloading)
    -   [Interaction with Generics](#interaction-with-generics)
    -   [Interaction with `unused`](#interaction-with-unused)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)

<!-- tocstop -->

## Abstract

Here we discuss the addition of default values for function arguments in Carbon, along
with limited interoperability support for the same in C++. The proposed Carbon syntax and semantics
closely emulate C++. We finish with some discussion for future work with general
support for defaults in other pattern-matching contexts such as local variables and `match`

> alternatives.

Reviewer: (minor)

FWIW, I think you can be a bit more imperative. Maybe along the lines of:

```suggestion
Proposes adding default values for function arguments to Carbon, along
with limited interoperability support for the same in C++. The proposed syntax and semantics
closely emulate C++.

Also outlines future work to provide more general
support for defaults in other pattern-matching contexts such as local variables and `match`
alternatives.
```

Author: Done.

## Problem

When interoperating with C++ code, much of that code will use default arguments for functions, for
example when providing default values for `a` and `b` in constructor arguments:

```cpp
class C {
 public:
  C(int a = 0, int b = 0): a_(a), b_(b) {}
 private:
  int a_;
  int b_;
};
```

In projects that are adopting Carbon, developers need to be able to write Carbon objects that
fluently emulate the API of their C++ counterparts. This fluent emulation of C++ necessarily
includes the need for support for default argument values.

Carbon establishes named bindings for
[function parameters](/docs/design/functions.md#named-and-positional-parameters) using the
[_tuple-pattern_](/docs/design/pattern_matching.md#tuple-patterns) contained within an outer group
of parenthesis. During pattern matching Carbon converts the _tuple-pattern_ to a tuple extended

> type by way of [extended type decomposition](/docs/design/values.md#extended-type-conversions).

Reviewer 2: The scrutinee is what gets converted, not the pattern. In any event, I'm not sure this is relevant here; extended type decomposition is about how we destructure the scrutinee into its elements, but defaults only come into play after that destructuring has happened, so I don't think we need to discuss how it happened.

Same goes for the discussion of extended type decomposition in the next paragraph.

Author: Ok, makes sense. I've removed the discussion of extended types and flattened these two paragraphs into a single one.

## Background

### General Knowledge

Carbon's
[C++ interoperability goals](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
have bearing on this, as this feature is a necessary part of that overall goal.

In addition, readers should have a basic understanding of
[C++ default arguments](https://en.cppreference.com/cpp/language/default_arguments)
as this proposed Carbon design largely mirrors this.

### Vocabulary

For the scope of this document we define certain terms as follows.

-   **Parameters** are the bindings between names and values used in a function declaration,
    declared in Carbon using its pattern-matching system.
-   **Arguments** are the actual values passed by a caller of the function at runtime, and are the
    scrutinee during pattern matching against the function parameter pattern.
-   **Default Argument Values**, sometimes shortened to default arguments or even defaults,
    are values for function parameters defined in the function declaration that will be provided
    by the compiler at the function call site when those arguments are not provided by the
    developer.

> #### Syntactic, Semantic, and Binding Parameters

Reviewer 2: Rather than defining these different kinds of parameters, and then using those definitions in the rest of the doc, I would suggest using this section to pivot from talking about parameters to talking about patterns. As a sketch of what that could look like, this section could say:

- In Carbon, a function parameter list is actually a tuple pattern, and an argument list is actually a tuple literal expression. When you call a function, the argument tuple is matched against the parameter pattern.
- Like any tuple pattern, the parameter list can contain arbitrarily complex subpatterns, not just name bindings.
- So the design for defaults for function parameters, we need to design how they fit into the pattern syntax, and how they affect pattern matching.

Then, try to rephrase the doc from here down to focus primarily on patterns rather than parameters, especially when you're focusing on granular details like the syntactic/semantic distinction (which is really a top-level/not top-level distinction), so you shouldn't need the definitions in this section. I've added some suggestions below to illustrate what I mean.

I think that will also make it easier to clearly state some of the major design ideas and choices in this proposal, e.g.
- Defaults are attached to the elements of tuple and struct patterns.
- `var` patterns cannot contain defaults.
- Defaults are only permitted in function signatures (and this is a choice we're making, not something that's true by definition).

Author: This feedback is in the Vocabulary section, so I decided to pivot the document to patterns starting in the Proposal section immediately below. I also made sure the points discussed here are brought up with more clarity in the Problem statement. I agree that the distinction around syntactic, semantic, and binding parameters isn't adding value so I removed it. And I've taken a shot at focusing the document more on patterns. PTAL.

#### Discord Discussions

-   [2026-07-24](https://discord.com/channels/655572317891461132/748959784815951963/1530303233942229053)
    Some questions and answers around the match process for patterns with defaults, as well as what
    to do with implicit conversions and defaults in structures.

#### Leads Questions

The work on this proposal exposed a few questions for Carbon Leads, enumerated here:

-   [#7529](https://github.com/carbon-language/carbon-lang/issues/7529) Do nested patterns with
    entirely specified default values also need to specify a default value to accept callee elision?
-   [#7530](https://github.com/carbon-language/carbon-lang/issues/7530) Which de-structured name
    binding form should accept defaults?

## Proposal

We propose adding initial support for default values for function arguments to Carbon. This will
require modification to the [_tuple-pattern_](/docs/design/pattern_matching.md#tuple-patterns) and
[_struct-pattern_](/docs/design/pattern_matching.md#struct-patterns) elements of pattern matching.

Following any subpattern in either _struct-pattern_ or _tuple-pattern_, the programmer can declare

> a default value with the addition of an optional assignment symbol `=` followed by a required

Reviewer 2: ```suggestion
a default value with the addition of an assignment symbol `=` followed by a
```

Given that the programmer is declaring a default value, the `=` is not optional.

Author: Yeah I meant to say declaring a default was optional, repaired.

> _value-expression_.

Reviewer 2: This should mention the caveat that you cannot declare a default inside a `var` pattern.

Author: Done.

In addition, we propose adding limited interoperation support with C++ for specifying default
values for function arguments. Most of the syntax and rules are designed to mirror that of function
argument default values in C++, to ease "tech transfer" for C++ developers to Carbon as well as
possibly reduce interoperability implementation costs.

> For now, we limit default value expressions to `musteval`, meaning they must be able to evaluate

Reviewer 2: Since `musteval` hasn't been proposed yet, it might be simpler to avoid mentioning it here, and just say that it must be a concrete constant expression or something like that.

Author: Done

  return x;
}

fn Main() {
  Core.Print(DefaultThree());  // `3`
  Core.Print(DefaultThree(7));  // `7`
}
```

## Out of Scope

In order to expedite the implementation work of the C++ interoperability requirements, this
proposal is deliberately scoped to focus on default values for function arguments. What follows is
a non-exhaustive list of elements we will not consider for change here:

-   Function overloading
-   Initializer expressions that are not `musteval`
-   Default argument values that have `ref` bindings
-   Defaults for function positional parameters
-   Defaults for variadic functions
-   Implicit conversions during pattern matching
-   Default values for local bindings, for now the compiler will reject
-   Default values for `match` alternative arms, for now the compiler will reject

## Details

This proposal evolved out of the goal of providing interoperability with C++ code expecting default
values for function arguments, including with function overloading, and as such, we've designed the
proposed syntax changes to be largely unsurprising to C++ developers.

Both _struct-pattern_ and _tuple-pattern_ accept nesting of each other, and so both will need to be

> modified to accept defaults in this context.

Reviewer 2: I don't think it's true that they both _need to_ be modified; it would be perfectly feasible and consistent (and would address all our interop use cases) to only support defaulting in tuple patterns. Supporting struct patterns as well is a choice we're making.

Author: Good point, modified.

Like in C++, defaults can be specified in a Carbon function declaration or definition. Unlike in C++,
Carbon requires defaults to be specified in the first declaration. Subsequent declarations must
either repeat the defaults exactly or provide a `= _` in every place a default value was specified
in the first declaration, following the convention of `where _` for constants established by
[p1084](https://github.com/carbon-language/carbon-lang/pull/1084).

### Tuple Patterns

Function parameters are declared in a _tuple-pattern_ surrounded by parenthesis. Defaults may be
specified as an optional addition for any, all, or none of the named bindings in a tuple in the

> function parameters list, at any level of nesting. However, at any particular level of nesting,

Reviewer 2: As an example of rephrasing to focus on patterns rather than parameters:

```suggestion
Defaults may be specified as an optional addition for any, all, or none of the elements of
a tuple pattern, at any level of nesting. However, at any particular level of nesting,
```

Author: Done.

elided parameters in unspecified order. For example:

```carbon
fn TwoDefaults(x: i32 = 0, y: i32 = 1);

TwoDefaults();  // OK, x = 0, y = 1
TwoDefauts(1);  // OK, x = 1, y = 1
TwoDefaults(, 3);  // ERROR, value not specified for first argument
```

Note that the right-to-left elision matching process happens at the individual tuple level, meaning
that nested _tuple-pattern_ scrutinees may also elide values in right-to-left order without being
subject to the constraint of where they are nested in a higher-level containing _tuple-pattern_.
For example the following is valid:

```carbon
fn Nested(a: i32, (b: i32, c: i32 = 1), d: i32);

Nested(1, (2), 3);  // OK, a = 1, b = 2, c = 1, d = 3
```

#### Pattern Matching

During the arity check phase of tuple pattern matching, we can any missing elements to the
scrutinee, drawing from the extended type of the defaults provided in the function parameter
pattern. Any type or other match issues created by these additions will be caught by further stages
in the match process.

### Struct Patterns

It is possible to nest a _struct-pattern_ inside of a _tuple-pattern_ in function parameter lists,

> so we must also extend default values to name bindings inside of the _struct-pattern_. Since the

Reviewer 2: As another example of rephrasing to focus on patterns:

```suggestion
It is possible to nest a _struct-pattern_ inside of a _tuple-pattern_,
so we must also extend default values to elements of the _struct-pattern_. Since the
```

However, as with my comment earlier, I don't think it's true that we _must_ do this; we're _choosing_ to do it.

Author: Done.

For _struct-pattern_, the compiler will supply the extended type for default values provided by the
pattern for any member names that are missing from the scrutinee. As with _tuple-pattern_, we expect
the rest of the pattern matching process to proceed as designed.

### Nesting Parameter Defaults

Defaults supplied at different levels of nesting have different implications for pattern matching.
For example:

```carbon
// Accepts 0 or 1 tuple argument. The supplied tuple can have 0-2 elements, the struct in the
// second position can have 0-2 members but they must be named `j` and/or `k`.
fn Binding((i: i32 = 1, {j: i32 = 2, k: i32 = 3}));

// Accepts 0 or 1 tuple argument. The supplied tuple can have 0-2 elements. The supplied struct
// must have a `j` and `k` member.
fn Semantic((i: i32 = 1, {j: i32, k: i32} = {.j = 2, .k = 3}));

// Accepts 0 or 1 tuple argument. If the tuple is supplied it must be fully-formed, and the struct
// in the second position must provide a `j` and `k` member and have no other members.
fn Syntactic((i: i32, {j: i32, k: i32}) = (1, {.j = 2, .k = 3}));
```

We require that at most one default be provided for each binding, and will reject code providing

> multiple defaults for a name binding.

Reviewer 2: I don't think this should be tied to bindings. For example, I think we want to disallow this, even though no binding has more than one default, because there are two defaults for `.b.x`:

```
fn F({a: i32, .b = {_} = {.x = 1}} = {.a = 1, .b = {.x = 2}})
```

On the other hand, we might very well want to allow something like

```
fn G({a: i32 = 0, b: i32 = 1} = {})
```

Maybe the rule should be something like "at most one default for each primitive struct or tuple pattern element", where "primitive" means that it's not a struct or tuple pattern itself?

Author: Ooh, much better put, stealing that, thanks!

### Interaction with Function Overloading

We acknowledge that support for default values, with the accompanying changes in the pattern
matching system, is adding a form of function overloading support to Carbon. However, the function
overloading system in Carbon is under active design and development at this time. Also, we are
intentionally limiting the scope of this proposal to default values for function parameters for the
sake of expedience. Therefore our hope is that this proposal is minimally disruptive to the overall
effort.

> We will add further content here as needed to clarify the exact relationship between these two

Reviewer 2: We don't make substantive changes to proposal docs after they're adopted (we just write new proposals to extend or supersede them), so this offer will need to be removed before the PR is submitted. That being the case, it seems simpler and safer to remove it now; any meta-discussion about how this proposal will evolve during review will probably be clearer outside the proposal document anyway.

Author: Done.

For future work, we may want to consider the use case in C++ for unused variables in a function
declaration whose default initializers have side effects intended to run in the callee scope.

## Rationale

This proposal advances the following Carbon goals:

-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code):
    Default values for function arguments are a widely-used feature of C++, so Carbon needs some
    utility to express these to support interop.
-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write):
    Allowing default values can increase the brevity and clarity of the calling code, by only
    specifying the arguments that are meaningful in the context of the call.
-   [Fast and scalable development](/docs/project/goals.md#fast-and-scalable-development): Support
    for default values allows some API changes to occur when adding arguments to existing methods,
    requiring no changes to existing client code beyond recompilation.

> ## Alternatives considered

Reviewer 2: Somewhere we should give the rationale for supporting defaults on struct patterns (despite not needing them for interop) and for disallowing defaults in local patterns. Listing those as alternatives considered might be the best vehicle for that.

Author: Added both.

**TBD** leads questions resolution, further feedback on this proposal.

> -   We considered an implementation of defaults only at the highest level of nesting in a

Reviewer 2: Typically each alternative gets its own subsection, to make it easier to cite.

Author: Done.
